x-pack/plugins/security/server/authentication/authenticator.ts

```diff
@@ -0,0 +1,388 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import {
+  SessionStorageFactory,
+  SessionStorage,
+  KibanaRequest,
+  LoggerFactory,
+  Logger,
+  HttpServiceSetup,
+  ClusterClient,
+} from '../../../../../src/core/server';
+import { ConfigType } from '../config';
+import { getErrorStatusCode } from '../errors';
+
+import {
+  AuthenticationProviderOptions,
+  BaseAuthenticationProvider,
+  BasicAuthenticationProvider,
+  KerberosAuthenticationProvider,
+  SAMLAuthenticationProvider,
+  TokenAuthenticationProvider,
+  OIDCAuthenticationProvider,
+} from './providers';
+import { AuthenticationResult } from './authentication_result';
+import { DeauthenticationResult } from './deauthentication_result';
+import { AuthenticationProviderSpecificOptions } from './providers/base';
+import { Tokens } from './tokens';
+
+/**
+ * The shape of the session that is actually stored in the cookie.
+ */
+export interface ProviderSession {
+  provider: string;
```

Comment:
I plan to expose `registerProvider` as a part of security setup contract at some point so that 3rd-party plugins could extend authc, and hence `provider` can be any string